core-bundle/src/Security/TwoFactor/TrustedDevice/TrustedDeviceManager.php

```diff
@@ -0,0 +1,154 @@
+        $this->trustedTokenStorage = $trustedTokenStorage;
+        $this->entityManager = $entityManager;
+        $this->httpClient = $httpClient;
+    }
+
+    public function addTrustedDevice($user, string $firewallName): void
+    {
+        if (!$user instanceof User || !$user instanceof TrustedDeviceInterface) {
+            return;
+        }
+
+        $username = $user->getUsername();
+        $version = $this->getTrustedTokenVersion($user);
+
+        $userAgent = $this->requestStack->getMasterRequest()->headers->get('User-Agent');
+        $parser = Parser::create();
+        $parsedUserAgent = $parser->parse($userAgent);
+
+        $geolocation = $this->getGeoLocation();
+        $country = null;
+
+        if (\array_key_exists('country', $geolocation)) {
+            $country = $geolocation['country'];
+        }
+
+        $this->trustedTokenStorage->addTrustedToken($username, $firewallName, $version);
+
+        $trustedDevice = new TrustedDevice();
+        $trustedDevice
+            ->setUserId((int) $user->id)
+            ->setUserClass(\get_class($user))
+            ->setCreated(new \DateTime())
+            ->setCookieValue($this->trustedTokenStorage->getCookieValue())
+            ->setUserAgent($userAgent)
+            ->setUaFamily($parsedUserAgent->ua->family)
+            ->setOsFamily($parsedUserAgent->os->family)
+            ->setDeviceFamily($parsedUserAgent->device->family)
+            ->setVersion($version)
+            ->setCountry(strtolower($country))
+        ;
+
+        $this->entityManager->persist($trustedDevice);
+        $this->entityManager->flush();
+    }
+
+    public function isTrustedDevice($user, string $firewallName): bool
+    {
+        if (!($user instanceof User)) {
+            return false;
+        }
+
+        $username = $user->getUsername();
+        $version = $this->getTrustedTokenVersion($user);
+
+        return $this->trustedTokenStorage->hasTrustedToken($username, $firewallName, $version);
+    }
+
+    public function clearTrustedDevices(User $user): void
+    {
+        /** @var TrustedDeviceRepository $trustedDeviceRepository */
+        $trustedDeviceRepository = $this->entityManager->getRepository(TrustedDevice::class);
+        $trustedDevices = $trustedDeviceRepository->findForUser($user);
+
+        foreach ($trustedDevices as $trustedDevice) {
+            $this->entityManager->remove($trustedDevice);
+        }
+
+        $this->entityManager->flush();
+
+        ++$user->trustedVersion;
+        $user->save();
+    }
+
+    private function getTrustedTokenVersion($user): int
+    {
+        if ($user instanceof TrustedDeviceInterface) {
+            return $user->getTrustedTokenVersion();
+        }
+
+        return self::DEFAULT_TOKEN_VERSION;
+    }
+
+    private function getGeoLocation(): array
+    {
+        $geolocation = [];
+
+        try {
+            $response = $this->httpClient->request('GET', 'https://ipinfo.io/geo');
```

Comment:
Doesn't this request retrieve the location of the server and not of the client?

Furthermore, in view of the GDPR, it would not be allowed to send the client IP to a geolocation service.

Reply:
if thats really a problem, I can only think of an option to disable the location tracking?

Reply:
Shouldn't we use `$_SERVER['REMOTE_ADDR']` here?

Reply:
It is now configurable if geolocation should be used or not. 
And yes, it was the server IP 🙈. Fixed in 00b14fa0e5dcce8342a03c756d65a6d9f4e2c233.